Mathlib/Computability/NFA.lean

```diff
@@ -104,6 +152,150 @@ def accepts : Language α := {x | ∃ S ∈ M.accept, S ∈ M.eval x}
+      ⟨h ▸ Path.nil s⟩
+    | a :: x =>
+      have h : ∃ s' ∈ M.step s a, t ∈ M.evalFrom {s'} x :=
+        by rw [evalFrom_cons, mem_evalFrom_iff_exists, stepSet_singleton] at h; exact h
+      let ⟨s', h₁, h₂⟩ := h
+      let ⟨p'⟩ := Path.evalFrom_iff_nonempty.1 h₂
+      ⟨Path.cons s' _ _ _ _ h₁ p'⟩
+  mpr p := match p with
+    | ⟨nil s⟩ => by simp
+    | ⟨cons s' s t a x h₁ h₂⟩ => by
+      rw [evalFrom_cons, stepSet_singleton, mem_evalFrom_iff_exists]
+      exact ⟨s', h₁, Path.evalFrom_iff_nonempty.2 ⟨h₂⟩⟩
+
+theorem Path.accepts_iff_nonempty {x : List α} :
+    x ∈ M.accepts ↔ ∃ s ∈ M.start, ∃ t ∈ M.accept, Nonempty (M.Path s t x) := by
+  simp only [← Path.evalFrom_iff_nonempty, M.mem_accepts, M.mem_evalFrom_iff_exists (S := M.start)]
+  tauto
+
+/-- Append two paths. -/
+def Path.append {s t u : σ} {x y : List α} (p : M.Path s t x) (q : M.Path t u y) :
+    M.Path s u (x ++ y) :=
+  match p with
+  | nil _ => q
+  | cons s' _ _ _ _ h p' =>
+    cons s' _ _ _ _ h (p'.append q)
+
+/-- Split `p` at the point just after it has matched `x`. -/
+def Path.splitAfter {s u : σ} {y : List α} (x : List α) (p : M.Path s u (x ++ y)) :
+    (t : σ) × M.Path s t x × M.Path t u y :=
+  match x, p with
+  | [], p => ⟨s, nil s, p⟩
+  | _ :: x, cons s' _ _ _ _ h p =>
+    let ⟨t, p', q⟩ := p.splitAfter x
+    ⟨t, cons s' _ _ _ _ h p', q⟩
+
+/-- Split `p` at the intermediate state `t`. -/
+def Path.splitAt [DecidableEq σ] {x : List α} {s u : σ}
+    (p : M.Path s u x) (t : σ) (h : t ∈ p.supp) :
+    (x₁ x₂ : List α) × M.Path s t x₁ × M.Path t u x₂ ×' x₁ ++ x₂ = x :=
+  if e : s = t then
+    ⟨[], x, e ▸ nil s, e ▸ p, rfl⟩
+  else
+    match p with
+    | nil _ => by simp at h; tauto
+    | cons s' _ _ a x h' p =>
+      have ⟨x₁, x₂, p', q, e'⟩ := p.splitAt t (by simp at h; tauto)
+      ⟨a :: x₁, x₂, cons s' _ _ _ _ h' p', q, e' ▸ rfl⟩
+
+/-- Knowing that the length of the word guarantees that a state will repeat
+somewhere along the path, find one such state and split the path around it. -/
+def Path.splitAtRepeatingState [DecidableEq σ] {x : List α} {s u : σ}
+    (p : M.Path s u x) (h : p.supp.card ≤ x.length) :
+    (t : σ) × (x₁ x₂ x₃ : List α) × M.Path s t x₁ × M.Path t t x₂ × M.Path t u x₃ ×'
+    x = x₁ ++ x₂ ++ x₃ ∧ x₂ ≠ [] :=
+  match p with
+  | nil _ => by simp at h
+  | cons s₁ _ _ a x h₁ p =>
+    if h' : s ∈ p.supp then
+      have ⟨x₂, x₃, p₂, p₃, e⟩ := p.splitAt s h'
+      have p₂ : M.Path s s (a :: x₂) := cons s₁ _ _ _ _ h₁ p₂
+      ⟨s, [], a :: x₂, x₃, nil s, p₂, p₃, e ▸ rfl, by simp⟩
+    else
+      have hle : p.supp.card ≤ x.length := by simpa [h', add_comm 1] using h
+      have ⟨t, x₁, x₂, x₃, p₁, p₂, p₃, e, hne⟩ := p.splitAtRepeatingState hle
+      ⟨t, a :: x₁, x₂, x₃, cons s₁ _ _ _ _ h₁ p₁, p₂, p₃, e ▸ rfl, hne⟩
+
+/-- Repeat a looping path to construct a path for `x∗`. -/
+noncomputable def Path.ofMemKstar {x x' : List α} {s : σ} (p : M.Path s s x)
+    (h : x' ∈ ({x}∗ : Language α)) : M.Path s s x' := by
```

Comment:
Can you make `x'` explicit so that one can tell from the infoview what its value is? Currently, it looks like `Path.ofMemKstar p ⋯`